multimodal/health_multimodal/image/model/model.py

```diff
@@ -0,0 +1,167 @@
+        Args:
+            input_img: input tensor image [B, C, H, W]
+            normalize: If set to True, the embeddings are l2-normalized.
+        Returns:
+            projected_embeddings: tensor of embeddings in shape [batch, n_patches_h, n_patches_w, feature_size]
+        """
+        assert self.training is False, "This function is only implemented for evaluation mode"
+        outputs = self.forward(input_img)
+        projected_embeddings = outputs.projected_patch_embeddings.detach()  # type: ignore
+        if normalize:
+            projected_embeddings = F.normalize(projected_embeddings, dim=1)
+        projected_embeddings = projected_embeddings.permute([0, 2, 3, 1])  # B D H W -> B H W D (D: Features)
+        return projected_embeddings
+
+
+class ImageEncoder(nn.Module):
+    def __init__(self, img_model_type: str):
+        super().__init__()
+        self.img_model_type = img_model_type
+        self.encoder = self._create_encoder()
+
+    def _create_encoder(self, **kwargs: Any) -> nn.Module:
+        if self.img_model_type == 'resnet18':
+            encoder = resnet18(return_all_feature_maps=True, pretrained=True, **kwargs)
+        elif self.img_model_type == 'resnet50':
+            encoder = resnet50(return_all_feature_maps=True, pretrained=True, **kwargs)
+        else:
+            raise NotImplementedError
+
+        return encoder
+
+    def forward(self, x: torch.Tensor, return_patch_embeddings: bool = False) -> Union[T, Tuple[T, T]]:
+        x = self.encoder(x)
+        x = x[-1] if isinstance(x, list) else x
+        avg_pooled_emb = torch.flatten(torch.nn.functional.adaptive_avg_pool2d(x, (1, 1)), 1)
+        if return_patch_embeddings:
+            return x, avg_pooled_emb
+        else:
+            return avg_pooled_emb
+
+    def reload_encoder_with_dilation(self, replace_stride_with_dilation: List[bool] = [False, False, True]) -> None:
+        """
+        This is a workaround for enabling dilated convolutions after the model initialization.
+
+        :param replace_stride_with_dilation: for each layer to replace the 2x2 stride with a dilated convolution
+        """
+        if self.img_model_type == 'resnet18':
+            # resnet18 uses BasicBlock implementation, which does not support dilated convolutions. See resnets.py in pl_bolts.
+            raise NotImplementedError("resnet18 does not support dilated convolutions")
+
+        device = next(self.encoder.parameters()).device
+        new_encoder = self._create_encoder(replace_stride_with_dilation=replace_stride_with_dilation).to(device)
+
+        if self.encoder.training:
+            new_encoder.train()
+        else:
+            new_encoder.eval()
+
+        new_encoder.load_state_dict(self.encoder.state_dict())
+        self.encoder = new_encoder
+
+
+def get_encoder_output_dim(module: torch.nn.Module) -> int:
+    """
+    Calculates the output dimension of ssl encoder by making a single forward pass.
```

Comment:
It's not clear what this means. Maybe something like "Calculate the number of channels in the output of the SSL encoder [etc.]"